### PR TITLE
Adding documentation for StateHoldr

### PR DESCRIPTION
**Pull request by gpnoel**

### CODE REVIEW

File: Storage.md [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+
+```typescript
+ItemsHolder.displayContainer();
+```
+
+More can be read about ItemsHoldr on its [Readme](https://github.com/FullScreenShenanigans/ItemsHoldr/blob/master/README.md).
+
+# StateHoldr
+
+StateHoldr is a module for tracking changes of items in ItemsHolder in objects called collections.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: Can you explain what a collection is?

File: Storage.md [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+Like ItemsHoldr, StateHoldr has its own prefix property.
```
Comment (JoshuaKGoldberg) [inline Storage.md:222]: Can you elaborate on how the two work together / are separate?

File: Storage.md [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+```typescript
+let StateHolder = new StateHoldr({
+    ItemsHolder: new ItemsHoldr(),
+    prefix: "StateHolder"
+});
+```
+
+## Changes
+
+Changes are stored as key-value pairs under an item.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: > under an item

What is an item verses a collection (verses a change)?

File: Storage.md [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+To add a change to the current collection, use `addChange`.
+`addChange` takes in the key of the item, an attribute of the item, and the value being stored.
+
+```typescript
+StateHolder.addChange("car", "color", "red");
+```
+
+If the name of the collection is known, whether or not its the current collection, a change can be added with `addCollectionChange`.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: It would be good to mention before here that there is such a concept as a "current" collection.

File: Storage.md [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+
+```typescript
+StateHolder.addCollectionChange("other_collection", "car", "color", "red");
+```
+
+Changes for an item in the current collection can be copied to another object with `applyChanges`.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: I read this first as copying one item's changes in the collection to another. Can you rephrase to be more clear?

File: Storage.md [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+```typescript
+ItemsHolder.displayContainer();
+```
+
+More can be read about ItemsHoldr on its [Readme](https://github.com/FullScreenShenanigans/ItemsHoldr/blob/master/README.md).
+
+# StateHoldr
+
+StateHoldr is a module for tracking changes of items in ItemsHolder in objects called collections.
+A collection is an object whose keys are item names in ItemsHolder and whose values are key-value pairs of of item attributes and values.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: > whose keys are item names in ItemsHolder

I wrote StateHoldr and this slightly confuses me. What do you mean by key? What do you mean by value?

File: Storage.md [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+These key-value pairs of attributes and values are referred to as changes.
+Changes are separate from the actual values for items in ItemsHoldr.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: This is also vague. You're diving into technical details without explaining the top-level of how someone would view StateHoldr from the outside.

File: Storage.md [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+
+## Prefix
+
+Like ItemsHoldr, StateHoldr has its own prefix property.
+The prefixes for StateHoldr and ItemsHoldr are not the same and do not affect each other.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: > and do not affect each other

No. StateHoldr routes all storage through ItemsHoldr so by nature it uses the ItemsHoldr prefix as well.

File: Storage.md [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+```typescript
+let StateHolder = new StateHoldr({
+    ItemsHolder: new ItemsHoldr(),
+    prefix: "StateHolder"
+});
+```
+
+## Collections
+
+StateHoldr keeps track of the collections its recorded.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: Redundant sentence?

File: Storage.md [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+It is saved as `stateCollectionKeys` in ItemsHolder.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: ? What is "It" here?

File: Storage.md [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+StateHoldr always points to one collection, the current collection, at any time.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: What do you mean by "points to"?

File: Storage.md [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+## Changes
+
+To add a change to the current collection, use `addChange`.
+`addChange` takes in the key of the item, an attribute of the item, and the value being stored.
+
+```typescript
+StateHolder.addChange("car", "color", "red");
+```
+
+If the name of the collection is known, whether or not its the current collection, a change can be added with `addCollectionChange`.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: Grammar for here and other places: "its" is not the same as "it's". Check your usage.

File: Storage.md (hunk 6 above)
Comment (gpnoel) [inline Storage.md]: I was trying to explain the structure for collections and how its item names paired with their attribute and values.

File: Storage.md [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+```typescript
+ItemsHolder.displayContainer();
+```
+
+More can be read about ItemsHoldr on its [Readme](https://github.com/FullScreenShenanigans/ItemsHoldr/blob/master/README.md).
+
+# StateHoldr
+
+StateHoldr is a module for tracking changes of items in ItemsHolder in objects called collections.
+A collection is an object whose keys are item names in ItemsHolder and whose values are key-value pairs of item attributes and values.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: > whose keys are item names

I don't understand. Elaborate?

File: Storage.md [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+These key-value pairs of attributes and values are referred to as changes.
+Collections are like a person and changes are the physical traits describing that person.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: ? This sentence is very vague.

File: Storage.md [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+
+## Prefix
+
+Like ItemsHoldr, StateHoldr has its own prefix property.
+All collections in StateHoldr are stored in ItemsHoldr prepended with the StateHoldr prefix.
+So the collections are stored in `localStorage` prepended with the ItemsHoldr prefix as well.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: This third sentence is unnecessary.

File: Storage.md [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+```typescript
+let StateHolder = new StateHoldr({
+    ItemsHolder: new ItemsHoldr(),
+    prefix: "StateHolder"
+});
+```
+
+## Collections
+
+Collections allow for various attributes for a single item to be grouped together versus storing individual items.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: Why bother with this first sentence if you're explaining what they are above?

File: Storage.md [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+The collections the module has recorded are stored in a list keyed by `stateCollectionKeys` in ItemsHolder (no StateHoldr prefix prepended).
+StateHoldr always has one collection as its current collection.
+
+`setCollection` sets the current collection.
+
+```typescript
+// sets the collection name and the collection's value
```
Comment (JoshuaKGoldberg) [inline Storage.md]: Is this comment necessary?

File: Storage.md [hunk 18]
```diff
@@ -0,0 +1,304 @@
...
+StateHolder.setCollection("newCollection", {
+    car: {
+        color: "red"
+    }
+});
+
+// sets only the name for the collection
+StateHolder.setCollection("newCollection");
```
Comment (JoshuaKGoldberg) [inline Storage.md]: Are you saying that the collection will be set to a blank object? As per https://github.com/FullScreenShenanigans/StateHoldr/blob/master/src/StateHoldr.ts#L134 I'm not sure that's accurate.

File: Storage.md [hunk 19]
```diff
@@ -0,0 +1,304 @@
...
+```
+
+If only a name is given, the collection is set to a blank object.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: Are you saying that the collection will be changed to a blank object? As per https://github.com/FullScreenShenanigans/StateHoldr/blob/master/src/StateHoldr.ts#L134 I'm not sure that's accurate.

File: Storage.md [hunk 20]
```diff
@@ -0,0 +1,304 @@
...
+
+`saveCollection` saves the current collection to ItemsHolder.
```
Comment (JoshuaKGoldberg) [inline Storage.md:261]: This requires explanation somewhere that the items aren't sent to ItemsHolder until they're saved.

File: Storage.md [hunk 21]
```diff
@@ -0,0 +1,304 @@
...
+## Changes
+
+To add a change to the current collection, use `addChange`.
+`addChange` takes in the key of the item, an attribute of the item, and the value being stored.
+
+```typescript
+StateHolder.addChange("car", "color", "red");
+```
+
+If the name of the collection is known, whether or not it's the current collection, a change can be added with `addCollectionChange`.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: > If the name of the collection is known

Vague and unnecessary. It would be better to phrase this as changing in any collection by its name.

File: Storage.md (hunk 13 above)
Comment (gpnoel) [inline Storage.md]: That was me trying to explain changes were keyed by item names.
How about this to replace the rest of that section.

Changes are key-value pairs of attributes for items in ItemsHolder and values.
A collection is a group of these changes keyed by the item name.
Collections allow for various attributes for a single item to be grouped together versus storing individual items.

To be honest I'm not sure how to best go about explaining collections and changes

Comment (JoshuaKGoldberg) [inline Storage.md]: A collection describes how named items in a group have been changed. For each item in the group, the collection will store a key-value pair of attributes and new values.

For example, the `"garage"` collection that contains a `"car"` item with `"color"` and `"name"`  changes could be described as:

``` js
"garage": {
    "car": {
        "color": "blue"
    }
}
```

_(then in the rest of the docs, keep referring to the garage collection)_

File: Storage.md [hunk 22]
```diff
@@ -0,0 +1,307 @@
...
+
+`setCollection` sets the current collection.
+
+```typescript
+StateHolder.setCollection("newCollection", {
+    car: {
+        color: "red"
+    }
+});
+```
```
Comment (JoshuaKGoldberg) [inline Storage.md]: Can you mention that you setting newCollection later will go back to the car/color/red one?

File: Storage.md [hunk 23]
```diff
@@ -0,0 +1,307 @@
...
+
+More can be read about ItemsHoldr on its [Readme](https://github.com/FullScreenShenanigans/ItemsHoldr/blob/master/README.md).
+
+# StateHoldr
+
+StateHoldr is a module for tracking changes of items in ItemsHolder in objects called collections.
+Changes are key-value pairs of attributes for items and values.
+A collection is a group of these changes keyed by the respective item name.
+Collections describe how items have changed.
```
Comment (JoshuaKGoldberg) [inline Storage.md]: I'd prefer it if you used the explanation I gave in a comment. This one is good, but that one is more clear.

File: Storage.md [hunk 24]
```diff
@@ -0,0 +1,320 @@
...
+```
+
+More can be read about ItemsHoldr on its [Readme](https://github.com/FullScreenShenanigans/ItemsHoldr/blob/master/README.md).
+
+# StateHoldr
+
+StateHoldr is a module for tracking changes of items in ItemsHolder in objects called collections.
+Changes are key-value pairs of attributes for items and values.
+A collection describes how named items in a group have been changed. For each item in the group, the collection will store a key-value pair of attributes and new values.
+For example, the "garage" collection that contains a "car" item with "color" and "name" changes could be described as:
```
Comment (JoshuaKGoldberg) [inline Storage.md]: Nit: _a_ "garage"

File: Storage.md [hunk 25]
```diff
@@ -0,0 +1,320 @@
...
+
+```typescript
+let garage: { [i:string]: any } = {
+    car: {
+        color: "red"
+    }
+}
```
Comment (JoshuaKGoldberg) [inline Storage.md]: Nit: missing semicolon
